File: 875-koko-eating-bananas/875-koko-eating-bananas.cpp
class Solution {
public:
    
    bool isPossible(vector<int>& piles, int sp, int h){
        
        int total_time=0;
        
        for(int i=0; i<piles.size(); i++){
            
            int div = piles[i]/sp;
            
            total_time += div;
            
            if(piles[i]%sp != 0) total_time++;
        }
        
        return total_time <= h;
    }
    
    int minEatingSpeed(vector<int>& piles, int h) {
        
        int maxi=piles[0];
        
        for(auto val : piles){
            maxi = max(val, maxi);
        }
        
        if(h == piles.size()){
            return maxi;
        }
        
        int lo = 1;
        int hi = maxi;
        
        int ans=0;
        
        while(lo <= hi){
            
            int mid = lo + (hi-lo)/2;
            
            if(isPossible(piles, mid, h) == true){
                ans = mid;
                hi = mid-1;
            }
            else{
                lo = mid+1;
            }
        }
        
        return ans;
        
    }
};

File: 875-koko-eating-bananas/875-koko-eating-bananas.py
import sys
class Solution:
    def canEat(self, piles: List[int], h: int, mid: int) -> bool:
        
        no_of_hours=0
        
        for pile in piles:
            hours = pile//mid
            no_of_hours += hours
            
            if pile%mid != 0:
                no_of_hours += 1
        
        return no_of_hours <= h
    
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        
        start = 1
        end = max(piles)
        
        ans = end
        
        while start <= end:
            mid = (start + end)//2
            
            if self.canEat(piles, h, mid) == True:
                ans = min(ans, mid)
                end = mid - 1
            
            else:
                start = mid + 1
        return ans